kep/61-real-load-aware-scheduling/README.md

```diff
@@ -0,0 +1,572 @@
+
+
+### Risks and Mitigations
+
+If utilization metrics are not available for a long time, we will fall back to the best fit bin pack based on allocations. There is no user action needed for this.
+
+
+To achieve X% utilization, it is recommended to set the value as X - 10 in practice. Refer to BestFitBinPack Score Plugin below for more details.
+
+## Design Details
+
+Our design consists of the following components as outlined in the diagram and described below. We propose two plugins namely "BestFitBinPack" and "SafeBalancing". Both of them use metrics from load watcher for scoring nodes with different algorithms.
+
+<img src="images/design.png" alt="design" width="720" height="628"/>
+
+
+
+### Metrics Provider
+
+This service provides metrics backed by a time-series database — for example, Prometheus, InfluxDB, [Kubernetes Metrics Server](https://github.com/kubernetes-sigs/metrics-server) etc.
+
+
+### Load Watcher/Load Analyser
+
+The load watcher and analyzer both run in a single process. The watcher is responsible for retrieving cluster-wide resource usage metrics like CPU, memory, network, and IO stats over windows of a specified duration from metrics providers above. It stores these in its local cache and persists several aggregations in the host local DB for fault tolerance. The analyzer is responsible for the detection of bad metrics and any remediation. Bad metrics could be due to missing metrics or metrics with considerable errors, making them anomalies. It can also be extended in the future to use ML models for analyzing metrics.
+
+Load watcher would cache metrics in the last 15-minute, 10-minute, and 5-minute windows, which can be queried via REST API exposed. A generic interface will be provided that can support fetching metrics from any metrics provider. Example is provided in Appendix.
+
+
+### DB
+
+This is a localhost database stored as a file. This DB aims to populate the load watcher's cache quickly if it is lost due to a crash and perform scheduling decisions without any impact on latency.
+The file will be stored in host file system, so it will be persisted across pod failures on the same node. For HA setups, each load watcher process will maintain its corresponding file locally.
+
+
+### Scheduler Plugins
+
+This uses the scheduler framework of K8s to incorporate our customized real load aware scheduler plugins without modifying the core scheduler code. The plugins we proposed mainly include the following two.
+
+- BestFitBinPack Plugin: It is best fit variant of bin pack algorithm that scores nodes by their actual resource utilization in a way that all utilized nodes have around x% of utilization.
+- Safe Balancing Plugin: It is a node sorting plugin that sorts nodes base on both the mean and the standard deviation of node resource utilization. It aims to balance not only the average load but also the risk caused by load variations.
+
+
+### BestFitBinPack Plugin
```

Comment:
For whatever reason, I can't comment in the old thread.

I would name it `TargetNodeLoad`

Reply:
Target is a bit misleading because we miss the target if we have no free nodes. Also achieving "Target load" is just a constraint it is not main goal, main goal is to increase cluster utilisation.
How about `LoadAwareNodePacking`? We always pack the nodes, whether below target or above, and it is load aware.

Reply:
> Also achieving "Target load" is just a constraint it is not main goal, main goal is to increase cluster utilisation.

I disagree. If that was your goal, you shouldn't have the penalty once the utilization is over 50. This plugin is trying to have nodes in the same target utilization, and if there are not enough resources left, it won't be able to. But that happens for any other Plugin.

@damemi @Huang-Wei moving the conversation of naming here, because github UI is great :)

Reply:
Okay your point is fair. I see three candidates:

1) `LoadAwareBinPack`
2) `LoadAwareBestFitBinPack`
3) `TargetNodeLoad`

My vote is for 1 or 2. One thing missing from 3 for me is bin pack behavior. It being a standard algorithm helps in user adoption for me just by looking at name.

Reply:
Voting for 2. `LoadAwareBestFitBinPack`

Reply:
Updated to `TargetLoadPacking` per discussion.